### PR TITLE
Number option support

### PR DESCRIPTION
**Pull request by nanasi880**
#56

Number option support.

スクリーンショット 2020-08-28 23 00 00

### CODE REVIEW

**Comment by toshimaru** [comment]
@nanasi880 Thanks for your pull request and I totally missed your Pull Request. 🙇 
I'm gonna review this within a week ❗

File: main.go [hunk 1]
```diff
@@ -132,3 +146,54 @@ func printThemes(cmd *cobra.Command) {
...
+	buf         []byte
+}
+
+func (w *numberWriter) Write(p []byte) (n int, err error) {
+
+	var (
+		original = p
+		tokenLen int
+	)
+	for i, c := range original {
```
Comment (toshimaru) [inline main.go:162]: [ask]Is it better to check if it continats `\n` before loop ❓ 

```suggestion
if bytes.Contains(p, []byte(string('\n'))) {
	for i, c := range original {
	   ...
```

File: main_test.go [hunk 2]
```diff
@@ -15,7 +15,7 @@ const (
 	unhighlightedGoCode = "[38;5;231mpackage main[0m[38;5;231m"
 )
 
-func TestMain(t *testing.T) {
+func TestMain(m *testing.M) {
```
Comment (toshimaru) [inline main_test.go:18]: Thanks 👍

File: main.go [hunk 3]
```diff
@@ -132,3 +146,54 @@ func printThemes(cmd *cobra.Command) {
...
+	}
+	return len(original), nil
+}
+
+func (w *numberWriter) Flush() error {
+	format := "%6d\t%s"
+	if w.currentLine > 999999 {
+		format = "%d\t%s"
+	}
+	_, err := fmt.Fprintf(w.w, format, w.currentLine, string(w.buf))
```
Comment (toshimaru) [inline main.go:196]: [ask]Can you remove the newline at EOF ❓ 

toshimaru_toshimbp___src_github_com_toshimaru_nyan

**Review by toshimaru — COMMENTED** [review]
Sorry for my super-late review. 🙇 

I'm new to using `io.Writer`, so I might be missing the point.  if so, let me know... 🙏 

And could you rebase the code ❓

**Comment by nanasi880** [comment]
Oh sorry. I missed your reply.
I'm gonna check your reply now.

File: main.go [hunk 4]
```diff
@@ -132,3 +146,54 @@ func printThemes(cmd *cobra.Command) {
 		cmd.Println()
 	}
 }
+
+type numberWriter struct {
+	w           io.Writer
+	currentLine int64
```
[the remark this answers is not in the document]
Comment (nanasi880) [reply main.go:152]: OK

File: main.go [hunk 5]
```diff
@@ -132,3 +146,54 @@ func printThemes(cmd *cobra.Command) {
...
+	buf         []byte
+}
+
+func (w *numberWriter) Write(p []byte) (n int, err error) {
+
+	var (
+		original = p
+		tokenLen int
```
[the remark this answers is not in the document]
Comment (nanasi880) [reply main.go:160]: In Go, length is conventionally an int (e.g., the return value of len("abc") is of type int)
I used to use the int type for that, but it's not required, so I'll change it to uint. :)

File: main.go [hunk 6]
```diff
@@ -132,3 +146,54 @@ func printThemes(cmd *cobra.Command) {
...
 }
+
+type numberWriter struct {
+	w           io.Writer
+	currentLine int64
+	buf         []byte
+}
+
+func (w *numberWriter) Write(p []byte) (n int, err error) {
+
```
[the remark this answers is not in the document]
Comment (nanasi880) [reply main.go:157]: OK

File: main.go [hunk 7]
```diff
@@ -100,15 +103,26 @@ func cmdMain(cmd *cobra.Command, args []string) (err error) {
 }
 
 func printData(data *[]byte, cmd *cobra.Command, lexer chroma.Lexer) {
+
```
[the remark this answers is not in the document]
Comment (nanasi880) [reply main.go:106]: OK

File: main.go (hunk 3 above)
Comment (nanasi880) [reply main.go:196]: This file is go.mod?
I checked the file, but the file ends with a newline to begin with. (i.e. the data entered has an EOF newline in the first place)
I do not think this should be trimmed on its own.

Comment (nanasi880) [reply main.go:196]: Oh, there are no EOF line breaks in the file, that's my mistake in checking. I'll think of a solution.

File: main.go (hunk 1 above)
Comment (nanasi880) [reply main.go:162]: I'll use "strings.Split" to rewrite it a bit more elegantly

**Comment by nanasi880** [comment]
> And could you rebase the code and add test code❓

OK, I will rebase the code.
The test code is difficult to know how to test (how to make sure it's colored), so do you have any ideas?

**Comment by nanasi880** [comment]
@toshimaru 

All points have been addressed except for the test code.
Can you review this content for now?

**Comment by nanasi880** [comment]
@toshimaru

I've added the test code.
About TestMain, TestMain is always run before every test.
The original TestMain calls the main function after adding `help`, which always had os.Exit(1) running.
In this state, the process is terminated before the test code (TestXXX function) is executed, so TestMain was removed. (TestMain is definitely broken.)

File: main.go (hunk 1 above)
Comment (nanasi880) [reply main.go:168]: I realized that I couldn't fix it that elegantly by using "bytes.Split".
I've added early returns for clarity, but note that adding "bytes.Contains" has increased the amount of slice `p` scanning.
Original: The scan amount is always O(N)
New implementation: scan volume is at worst O(2N) (if the data does not contain line breaks)

**Review by toshimaru — APPROVED** [review]
Hey, @nanasi880, very sorry to have kept you waiting. 🙇 
This is a great work, I'll merge this. 👍 (I'll fix fmt issue)

**Pull request merged**